arangod/Scheduler/WorkStealingThreadPool.cpp

```diff
@@ -0,0 +1,451 @@
+      }
+    } else if (stealAttempts > maxStealAttempts) {
+      // nothing to work -> go to sleep
+      goToSleep(guard);
+      if (maxStealAttempts > 2) {
+        maxStealAttempts /= 2;
+      }
+      stealAttempts = 0;
+    } else {
+      guard.unlock();
+      if (stealAttempts == 0) {
+        // try to insert ourselves in the hint chain
+        // threads with lower IDs spin longer, so we prefer to have them at the
+        // front of the chain, provided they are not sleeping already.
+        // We simply limit ourselves to 4 steps before giving up (note that the
+        // chain can contain loops)
+        auto* hint = &pool.hint;
+        for (unsigned i = 0; i < 4; ++i) {
+          auto h = hint->load(std::memory_order_relaxed);
+          auto& ts = pool._threadStates[h];
+          if (h > id || ts->sleeping.load(std::memory_order_relaxed)) {
+            prevHint.store(h, std::memory_order_relaxed);
+            hint->store(id, std::memory_order_relaxed);
+            break;
+          }
+          hint = &ts->prevHint;
+        }
+      }
+      auto work = stealWork();
+      if (work) {
+        doWork(*work);
+      } else {
+        basics::cpu_relax();
+        ++stealAttempts;
+      }
+    }
+  }
+}
+
+void WorkStealingThreadPool::ThreadState::goToSleep(
+    std::unique_lock<std::mutex>& guard) noexcept {
+  sleeping.store(true, std::memory_order_seq_cst);
+  // the store and load both need to be seq_cst to ensure proper ordering with
+  // the push that wakes us up
+  if (pushQueue.load(std::memory_order_seq_cst) == nullptr) {
+    pool.lastSleepIdx.store(id, std::memory_order_relaxed);
+    pool.numSleeping.fetch_add(1, std::memory_order_relaxed);
+    cv.wait_for(guard, std::chrono::milliseconds{100});
+    pool.numSleeping.fetch_sub(1, std::memory_order_relaxed);
+  }
+  sleeping.store(false, std::memory_order_relaxed);
+}
+
+auto WorkStealingThreadPool::ThreadState::pushMany(WorkItem* item)
+    -> std::unique_ptr<WorkItem> {
+  TRI_ASSERT(item != nullptr);
+  while (item->next != nullptr) {
+    auto next = item->next;
+    item->next = nullptr;
+    _queue.push_front(std::unique_ptr<WorkItem>(item));
+    item = next;
+    TRI_ASSERT(item != nullptr);
+  }
+  queueSize.store(_queue.size(), std::memory_order_relaxed);
+  TRI_ASSERT(item->next == nullptr);
+  return std::unique_ptr<WorkItem>{item};
+}
+
+void WorkStealingThreadPool::ThreadState::runWork(WorkItem& work) noexcept {
+  incCounter(pool._metrics.jobsDequeued);
+  if (pool._metrics.dequeueTimes) {
+    auto queueTime = std::chrono::duration<double>(
+        std::chrono::steady_clock::now() - work.enqueueTime);
+    auto d =
+        std::chrono::duration_cast<std::chrono::microseconds, double>(queueTime)
+            .count();
+    pool._metrics.dequeueTimes->count(d);
+  }
+  pool.statistics.dequeued.fetch_add(1, std::memory_order_relaxed);
+  try {
+    work.invoke();
+  } catch (...) {
+    LOG_TOPIC("d5fb3", WARN, Logger::FIXME)
```

Comment:
I would prefer catching all `std::exception`s here and then logging their proper exception message. Otherwise it is hard to figure out what exactly went wrong.